### ISSUE TITLE
#if and undefined variables

### ISSUE THREAD

**Issue by tmilker** [ractivejs/ractive#3284]
## Description:
I'm attempting to upgrade an app from 0.8-edge(commit: d35141c6b3) to 1.0.1 and running into an issue with #if and undefined variables.  I'm using a 'shown' variable in my template to control whether an in-template partial is displayed or not.  The 'shown' variable is not defined in the data until it is set to true when the element is clicked and expanded to be displayed.  The partial is wrapped in an #if because there may be hundreds of images to load and I don't want them all loaded on page load and instead wait until the first time the user clicks.  In the version of 0.8-edge I'm using, this works fine.  In 1.0.1, when 'shown' is set to true, the partial is not rendered.  I have some other filtering going on and if the filter changes and then is changed back to the selection that was set to true, then the partial is rendered and the images are loaded correctly.

## Versions affected:
1.0.1

## Reproduction:
http://jsfiddle.net/w5Leyo9v/2/
I've created a small example that recreates a similar situation to my app.  In 0.8-edge(commit above) it works.  In 1.0.1 if you leave 'shown' in data commented out, nothing happens when you click the buttons.  But if you uncomment 'shown' in data and re-run the fiddle, the buttons work fine.

## Workaround:
Make sure variable is defined before template is rendered.  This is kind of annoying in my situation since I'm combining data from different sources and have to make sure to put this in the correct spot.  If data was merged in later, it would also have to be accounted for.

**Comment by ceremcem** [comment]
I can confirm that only ~partial rendering~ *each loop* suffers from this issue. 

TL;DR;

1. Toggle list with `show` and `hide` buttons
2. See the list is shown and hidden
3. Comment out `shown: false` line inside `data`: 

```js
    template: '#dataTemplate',
    data: {
      //shown: false,
      BenchImage: [
        {
```

4. See the buttons are still setting `shown` variable but  ~`{{> benchImages}}` partial doesn't render~ *nothing inside `{{#each}}` loop is rendered*.

**Comment by tmilker** [comment]
It doesn't seem to be just partial rendering because if you replace the partial in the example with the content of the in-template partial, it still fails to work.

**Comment by evs-chris** [comment]
Your workaround is one of the two ways to address this issue. The other is to use explicit references e.g. `~/shown` rather than just `shown`. The reason for that is ractive no longer tries to wait to see if a particular reference happens to appear somewhere in the context stack before binding it, and instead immediately resolves ambiguous references in the immediate context, which is correct in _most_ (or at least many) cases.

If one of your iterated items happened to contain a `shown` member, it could also cause issues, so we now recommend remaining explicit with references wherever possible.

**Comment by ceremcem** [comment]
@tmilker We are both partially wrong and partially correct: The `{{#if shown}}` part doesn't work inside `{{#each}}` but it works outside of `{{#each}}`: http://jsfiddle.net/ceremcem/uqnL59hj/

**Comment by tmilker** [comment]
@evs-chris That workaround doesn't work in my real app.  How do I bind it correctly when the keypath is more like: 'data.0.tests.1.shown' and shown isn't defined until click?  I can find uses of ~/ but can't find any reference in the docs describing what ~/ means.

**Comment by evs-chris** [comment]
There are a handful of [other prefixes](https://ractive.js.org/api/#keypath-prefixes) for references that let you explicitly reference keypaths. So if your context is current `data.0.test.1.things.5` and you want to reference `data.0.test.1.shown`, you would use `../../shown` for pop `5` and `things` from the current keypath and add on `shown`. It works like filesystem path traversal, and it's also available since way back (pre 0.7) if you need to test it out in your current app.

**Comment by tmilker** [comment]
@evs-chris Thanks, totally get what was going on then and what is happening now.
